pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+			defaultCIDRConfig.Spec.PerNodeHostBits = ipv6PerNodeHostBits
+		}
+	}
+
+	existingConfigList.Items = append(existingConfigList.Items, *defaultCIDRConfig)
+
+	return
+}
+
+// reconcileCreate handles create ClusterCIDR events.
+func (r *multiCIDRRangeAllocator) reconcileCreate(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if needToAddFinalizer(clusterCIDR, clusterCIDRFinalizer) {
+		klog.V(3).Infof("Creating ClusterCIDR %s", clusterCIDR.Name)
+		if err := r.createClusterCIDR(clusterCIDR, false); err != nil {
+			klog.Errorf("Unable to create ClusterCIDR %s : %v", clusterCIDR.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// reconcileBootstrap handles creation of existing ClusterCIDRs.
+// adds a finalizer if not already present.
+func (r *multiCIDRRangeAllocator) reconcileBootstrap(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	terminating := false
+	// Create the ClusterCIDR only if the Spec has not been modified.
+	if clusterCIDR.Generation > 1 {
+		terminating = true
+		err := fmt.Errorf("CIDRs from ClusterCIDR %s will not be used for allocation as it was modified", clusterCIDR.Name)
+		klog.Errorf("ClusterCIDR Modified: %v", err)
+	}
+
+	klog.V(2).Infof("Creating ClusterCIDR %s during bootstrap", clusterCIDR.Name)
+	if err := r.createClusterCIDR(clusterCIDR, terminating); err != nil {
+		klog.Errorf("Unable to create ClusterCIDR %s: %v", clusterCIDR.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// createClusterCIDR creates and maps the cidrSets in the cidrMap.
+func (r *multiCIDRRangeAllocator) createClusterCIDR(clusterCIDR *networkingv1alpha1.ClusterCIDR, terminating bool) error {
+	nodeSelector, err := r.nodeSelectorKey(clusterCIDR)
+	if err != nil {
+		return fmt.Errorf("unable to get labelSelector key: %w", err)
+	}
+
+	clusterCIDRSet, err := r.createClusterCIDRSet(clusterCIDR, terminating)
+	if err != nil {
+		return fmt.Errorf("invalid ClusterCIDR: %w", err)
+	}
+
+	if clusterCIDRSet.IPv4CIDRSet == nil && clusterCIDRSet.IPv6CIDRSet == nil {
+		return errors.New("invalid ClusterCIDR: must provide IPv4 and/or IPv6 config")
+	}
+
+	if err := r.mapClusterCIDRSet(r.cidrMap, nodeSelector, clusterCIDRSet); err != nil {
+		return fmt.Errorf("unable to map clusterCIDRSet: %w", err)
+	}
+
+	// Make a copy so we don't mutate the shared informer cache.
+	updatedClusterCIDR := clusterCIDR.DeepCopy()
+	if needToAddFinalizer(clusterCIDR, clusterCIDRFinalizer) {
+		updatedClusterCIDR.ObjectMeta.Finalizers = append(clusterCIDR.ObjectMeta.Finalizers, clusterCIDRFinalizer)
+	}
+
+	if updatedClusterCIDR.ResourceVersion == "" {
+		// Create is only used for creating default ClusterCIDR.
+		if _, err := r.client.NetworkingV1alpha1().ClusterCIDRs().Create(context.TODO(), updatedClusterCIDR, metav1.CreateOptions{}); err != nil {
+			klog.V(2).Infof("Error creating ClusterCIDR %s: %v", clusterCIDR.Name, err)
+			return err
+		}
```

Comment:
what does this mean exactly? that we only use this for the one we are generating from flags? The one generated by `createDefaultClusterCIDR()` ?

if that is the case the match should be on the object name, the Resourceversion = "" is very hard to understand